Standardize password hash function usage

Joey3000 opened the pull request:
As preparation for introduction of a stronger **password** hash function in https://github.com/piwik/piwik/issues/5728, this standardizes the usage of the current password hash function across the Login and UsersManager plugins. Specifically:
- Use `UsersManager::getPasswordHash()` throughout, instead of previous direct use of `md5()` in some places.
- Concentrate the hash length sanity check in the newly created `UsersManager::checkPasswordHash()`. Previously, `UsersManager\API` and `Login\PasswordResetter` classes did it separately. Both checks were moved into the newly created function in the `UsersManager` class, as that is also where other `public static` password check functions reside.
- Replace the "md5" string with "hash" in affected variable names and comments.

No functional change is intended.

~~One little caveat is that I can't test at the moment. It would be great if someone could take this on a test ride.~~

---

To summarize the hash function uses by the Login and UsersManager plugins:
- The **password** hash function is now centralized in (https://github.com/piwik/piwik/blob/c54fc9e4e6072e7bffbb1a02b1c11645c50038e4/plugins/UsersManager/UsersManager.php#L153):
  
  ``` php
  public static function getPasswordHash($password)
  {
      // if change here, should also edit the installation process
      // to change how the root pwd is saved in the config file
      return md5($password);
  }
  ```

Following hash function uses were not modified:
- **`token_auth`** hash function is controlled by (https://github.com/piwik/piwik/blob/c54fc9e4e6072e7bffbb1a02b1c11645c50038e4/plugins/UsersManager/API.php#L789) (apparently for DB storage - this could be something for https://github.com/piwik/piwik/issues/9457):
  
  ``` php
  public function getTokenAuth($userLogin, $passwordHash)
  {
      UsersManager::checkPasswordHash($passwordHash, Piwik::translate('UsersManager_ExceptionPasswordMD5HashExpected'))
  
      return md5($userLogin . $passwordHash);
  }
  ```
  
    As well as by (https://github.com/Joey3000/piwik/blob/c54fc9e4e6072e7bffbb1a02b1c11645c50038e4/plugins/Login/SessionInitializer.php#L235) (apparently for cookie setting - this could be changed, e.g, at the same time as https://github.com/piwik/piwik/issues/9457 gets implemented, to have `token_auth` only change once):
  
  ``` php
  public static function getHashTokenAuth($login, $token_auth)
  {
      return md5($login . $token_auth);
  }
  ```
- **Password reset token** (per e-mail) is controlled by (https://github.com/piwik/piwik/blob/c54fc9e4e6072e7bffbb1a02b1c11645c50038e4/plugins/Login/PasswordResetter.php#L320):
  
  ``` php
  protected function hashData($data)
  {
      return Common::hash($data);
  }
  ```

Joey3000 committed:
Standardize usage of **password** hash function

As preparation for introduction of a stronger hash function in https://github.com/piwik/piwik/issues/5728, this standardizes the usage of the current **password** hash function across the Login and UsersManager plugins. Specifically:

* Use UsersManager::getPasswordHash() throughout, instead of previous direct use of md5() in some places

* Concentrate the hash length sanity check in the newly created UsersManager::checkPasswordHash(). Previously, UsersManager\Auth and Login\PasswordResetter classes did it separately. Both checks were moved into the newly created UsersManager class function, as that is also where other public static password check functions reside.

* Replaced the "md5" string with "hash" in affected variable names and comments

No functional change is intended.

Joey3000 committed:
Comment update only

As preparation for Common::hash() changing its hash function some time in future

Joey3000 committed:
Revert hash calculation on comparison

For performance reasons

Joey3000 committed:
Bugfix (syntax)

Joey3000 committed:
Syntax error fix

Joey3000 commented:
Does anyone know why the test fails?

I've tested the patch with piwik-2.16.0-b1.zip (on a remote test server, updated from piwik-2.15.0-b14 via a complete delete and re-upload via FTP, keeping the DB tables) and don't get any errors with PHP 5.5. Tested:
* As super user: Login plugin: login, logout; UsersManager plugin: create a "view" user, change own and "view" user's password
* As the new "view" user: Login plugin: login, logout
* Not logged in: password change (was accepted - failed correctly: server has e-mailing disabled)

I don't get any `passwordHash` related errors.

-------------------------

The only error I see does not seem to be related: UsersManager: When trying to change own password as the "view" user (not tested as super user), I get following error:

`SQLSTATE[42S22]: Column not found: 1054 Unknown column 'use_12_hour_clock' in 'field list'`

(Without trying to change the clock format. And similarly for `'use_24_hour_clock' ` if I try to change it to 24h-format.)

But that error could be due to the update from the 2.15.0-b14 beta. I'd need to try updating from a released version some time.

Joey3000 commented:
I found it, thanks to https://builds-artifacts.piwik.org/piwik/piwik/master/17449/ (missing semicolon).

The remaining `passwordHash`-related issues are due to the function argument renaming in https://github.com/piwik/piwik/pull/9480/files#diff-243038712dace7532e3b1856dd63274eR789. (Including the API list screenshot - scroll down to the "Module UsersManager". So, it's expected. But I can revert the renaming, if desired.

Please let me know if it'd be better to do this in the 3.0 branch (or not at all).

tsteur commented on plugins/UsersManager/API.php:
```diff
@@ -783,16 +783,13 @@ private function isUserTheOnlyUserHavingSuperUserAccess($userLogin)
      * @return string
      */
-    public function getTokenAuth($userLogin, $md5Password)
+    public function getTokenAuth($userLogin, $passwordHash)
```
That's why tests fail as it breaks the API. Unfortunately we cannot rename this one although it would be good :( 

Users call it eg with `index.php?module=API&method=API.getTokenAuth&userLogin=foo&md5Password=bar...`. This would break all applications, libraries etc that use this method (eg the Piwik Mobile App but also many other libraries and apps).

Unfortunately we do not have a versioning in our API to handle such cases. The only thing that I could imagine would be making `$md5Password` optional, add `$passwordHash` optional and then checking in the method that at least one of them is set and if not throw an exception.

We'd need to document this in the developer changelog and deprecate `$md5Password` parameter. We'd probably never actually remove that parameter instead deprecate the whole API and replace it with a REST API in a few years

tsteur commented:
PR looks good! Thx for that. Only one issue with the failing test. Explained it in a code comment

Joey3000 committed:
Revert renaming of getTokenAuth API parameter

To prevent API breakage - see https://github.com/piwik/piwik/pull/9480#discussion-diff-49282923

Joey3000 committed:
Fix missing `use Piwik\Piwik`

See https://builds-artifacts.piwik.org/piwik/piwik/master/17471/processed/Login_formless_login.png

Joey3000 commented:
Thanks for the feedback! I reverted the renaming (adding a comment in https://github.com/piwik/piwik/pull/9480/files#diff-243038712dace7532e3b1856dd63274eR786 instead) and fixed another issue left over.

I don't know what the remaining test failures are about, though. E.g.:
```
Piwik\Tests\System\PeriodIsRangeDateIsLastNMetadataAndNormalAPITest: Differences with expected in '/home/travis/build/piwik/piwik/tests/PHPUnit/Framework/TestCase/../../System/processed/test_periodIsRange_dateIsLastN_MetadataAndNormalAPI__Live.getVisitorProfile.xml'

Failed asserting that two DOM documents are equal.

--- Expected

+++ Actual

@@ @@

 <?xml version="1.0"?>

 <result>

 	<totalVisits>2</totalVisits>

 	<totalVisitDuration>361</totalVisitDuration>

 	<totalActions>2</totalActions>

+	<totalEvents>0</totalEvents>

+	<totalOutlinks>1</totalOutlinks>

+	<totalDownloads>0</totalDownloads>

 	<totalSearches>0</totalSearches>

+	<totalPageViews>1</totalPageViews>

[...]
```

As well as the moved text in the screenshots in https://builds-artifacts.piwik.org/piwik/piwik/master/17472/.

I would appreciate any hints.

tsteur commented:
Thx for the quick change! The failing test is not due to your change. I merged a couple of PRs yesterday but haven't fixed the build yet.

Joey3000 commented:
> The only error I see does not seem to be related: UsersManager: When trying to change own password as the "view" user (not tested as super user), I get following error:

  ```
  SQLSTATE[42S22]: Column not found: 1054 Unknown column 'use_12_hour_clock' in 'field list'
  ```

Reported in https://github.com/piwik/piwik/issues/9666.
